### PR TITLE
WIP #1517 next scrum

### PR DESCRIPTION
**Pull request by c1505**
Work in progress for #1517 

Added 2 scenarios that pass with the hotfix and fail without it.  The third on line 97 should be failing without the hotfix, but is passing.  Help would be appreciated.

### CODE REVIEW

**Comment by tansaku** [comment]
thanks @c1505 - will review shortly ...

**Comment by tansaku** [comment]
@c1505 running locally for me all of `features/hangouts/hangout.feature` passes, with one pending step:

```
Then(/^I should not see link "([^"]*)" with "([^"]*)"$/) do |arg1, arg2|
  pending # Write code here that turns the phrase above into concrete actions
end 
```

from 

```
 @time-travel-step
  Scenario: Don't show the Join live event link on the home page when event is not live # features/hangouts/hangout.feature:89
    Given the date is "2014/02/03 07:01:00 UTC"                                         # features/step_definitions/event_steps.rb:94
    When I am on the home page                                                          # features/step_definitions/projects_steps.rb:63
    Then I should not see "Scrum is live!"                                              # features/step_definitions/basic_steps.rb:182
    And I should not see link "Click to join!" with "http://hangout.test"               # features/hangouts/hangout.feature:93
      Undefined step: "I should not see link "Click to join!" with "http://hangout.test"" (Cucumber::Undefined)
      features/hangouts/hangout.feature:93:in `And I should not see link "Click to join!" with "http://hangout.test"'

```

**Comment by tansaku** [comment]
it does indeed pass with and without the hotfix for me.  I had a look by adding <% byebug %> into different partials, and using `And show me the page` in the cukes (CSS not loading - I'd love to fix that - thought I did that on LocalSupport - I think just set `Capybara.asset_host = 'http://localhost:3000'`).  I didn't immediately get to the bottom of the lack of failure.

I'm thinking that to make this more tractable we should move the specific wrapping of the bug into a separate feature file, so we can strip out the background steps that are actually necessary.

It's possible we might only see the issue when we have more than one repeating scrum.  If none of the above helps, then in principle we can pull down some production data to replicate.  It seemed to me that the situation that was leading to it was having two events, one of which had gone live.

Maybe that's the critical missing thing - you don't take any of the events live - you'll need something like:
```
    Given the Hangout for event "Scrum" has been started with details:
      | EventInstance link | http://hangout.test |
      | Started at         | 10:25:00 UTC        |

**Comment by tansaku** [comment]
@c1505 well done making further progress on this - are you feeling confident that the new tests will not fail appropriately?

I notice we are not testing the situation where the hangout actually gets taken live ...

I wonder if it might be a good idea to split a subset of cuke scenarios out into another feature file

I've noticed that the repeated setup for many scenarios in one feature file are causing our cukes to run slower - perhaps we could get two birds with one stone here by splitting out the scenarios for this into a separate feature file, and then making sure we nail the set of related circumstances of scrums being in their "live" time and actually being taken live (i.e. ping comes in from hangout plugin).

let me know what you think - feel free to hand off if you've gotten sick of working on this :-)

**Comment by c1505** [comment]
Not sick of this yet, but got physically sick last week so I haven't spent much time on this.  I think you are right that these new cuke scenarios should be in a separate file.  I will split these out and look again if I can figure out why the one scenario passes without the hotfix change you made.

**Comment by c1505** [comment]
Using "And show me the page", I can see the text that I expect even though it isn't styled appropriately.  The test does not see that text though.  Any ideas?

image

**Comment by tansaku** [comment]
@c1505 I think that might be because it's in a link, and the capybara `has_text` is looking for things in other html tags - it's a bit weird - but sometimes you have to explicitly check for the link text using `has_link` or `has_button` or what have you

**Comment by c1505** [comment]
It also fails with looking for just "is live".  If I change in capybara.rb to what you see below, cucumber can see the text.  I don't know why that much about cucumber including why it would be configured the way it is currently.  Curious to see what if anything breaks in CI so I'll just push it and see what happens.   
```ruby
# Capybara.javascript_driver = :poltergeist_billy
Capybara.javascript_driver = :webkit
```

**Comment by tansaku** [comment]
@c1505 that sounds like javascript dependencies - I'll have a closer look later today

**Comment by tansaku** [comment]
I'm running `be cucumber features/events/next_scrum.feature` locally, and that passes for me.  In CI a host of other things are failing ...

```
cucumber -p semaphoreci features/basic_layout.feature:47 # Scenario: Show Sponsors on desktop computer
cucumber -p semaphoreci features/basic_layout.feature:55 # Scenario: Hide Sponsors from Tablets
cucumber -p semaphoreci features/basic_layout.feature:63 # Scenario: Hide Sponsors from Smartphones
cucumber -p semaphoreci features/basic_layout.feature:71 # Scenario: Show Round banners on desktop computer
cucumber -p semaphoreci features/basic_layout.feature:77 # Scenario: Hide Round banners from Tablets
cucumber -p semaphoreci features/basic_layout.feature:83 # Scenario: Hide Round banners from Smartphones
cucumber -p semaphoreci features/documents.feature:82 # Scenario: Mercury editor shows Save and Cancel buttons, hides New Document button, Save button works
cucumber -p semaphoreci features/documents.feature:97 # Scenario: Mercury editor Cancel button works
cucumber -p semaphoreci features/documents.feature:130 # Scenario: A user can insert an image
cucumber -p semaphoreci features/documents.feature:143 # Scenario: Missing Image gets added when a user can inserts an invalid image
cucumber -p semaphoreci features/hire_me_modal.feature:12 # Scenario: Sending a message to user with 'Hire me' button
cucumber -p semaphoreci features/pages.feature:53 # Scenario: Mercury editor shows Save and Cancel buttons
cucumber -p semaphoreci features/pages.feature:60 # Scenario: Mercury editor Save button works
cucumber -p semaphoreci features/pages.feature:72 # Scenario: Mercury editor Cancel button works
cucumber -p semaphoreci features/projects/edit_project.feature:58 # Scenario: Editing Pitch content with Mercury Editor
```

because we are generally set up to rely on poltergeist.  Usually we should try and avoid adjusting capybara settings to fix things up.

I also notice the pull request includes some changes in other places, such as in `spec_helper` with:

```
ActiveRecord::Migration.maintain_test_schema!
```

while that's a change we might want to make, we need to be ultra-cautios about other minor changes creeping in when we are trying to track down the cause of strange behaviour

If I switch back to :poltergeist_billy I get the failure you mention:

```
  @javascript @time-travel-step
  Scenario: Show the next scrum join link on all pages # features/events/next_scrum.feature:47
    Given the date is "2014/02/03 07:01:00 UTC"        # features/step_definitions/event_steps.rb:94
    Given I am on the show page for event "Random"     # features/step_definitions/event_steps.rb:78
    Then I should see "Scrum is live!"                 # features/step_definitions/basic_steps.rb:184
      expected to find text "Scrum is live!" in "Log in Sign up ABOUT PROJECTS MEMBERS PREMIUM EVENTS GETTING STARTED Random Weekly retrospective Event type: PairProgramming Learn more about Scrums Learn More About Us Getting Started Dashboard Social Facebook Twitter Our Sponsors drie Craft Academy Mentive RubyMine Becoming a sponsor AgileVentures - Crowdsourced Learning Contact us Send a traditional email to info@agileventures.org." (RSpec::Expectations::ExpectationNotMetError)
      ./features/step_definitions/basic_steps.rb:186:in `/^I should( not)? see "([^"]*)"$/'
      features/events/next_scrum.feature:50:in `Then I should see "Scrum is live!"'

Failing Scenarios:
cucumber features/events/next_scrum.feature:47 # Scenario: Show the next scrum join link on all pages
```

To examine in more detail I adjusted the failing test like so:

```
  @javascript @time-travel-step
  Scenario: Show the next scrum join link on all pages
    Given the date is "2014/02/03 07:01:00 UTC"
    Given I am on the show page for event "Random"
    And I debug
    Then I should see "Scrum is live!"
```

and run just that one

```
be cucumber features/events/next_scrum.feature:47
```

That allows me to dump the html of the page and I see: 

```html
<a href=\"/events/scrum\">Scrum</a> is live!</div>
```

I then did some experimentation in the debugger:

```
(byebug) page.has_text? 'is live'
false
(byebug) page.body.has_text? 'is live'
*** NoMethodError Exception: undefined method `has_text?' for #<String:0x007fbbed1a8a00>
(byebug) page.body =~ /is live/
6346
(byebug) page.body.has_content? 'is live'
*** NoMethodError Exception: undefined method `has_content?' for #<String:0x007fbbeb4b8dc8>
```

Then I did some [googling](https://www.google.co.uk/search?q=capybara+has_text+not+finding+text) and skimmed the following:

https://github.com/teamcapybara/capybara/issues/844

http://stackoverflow.com/questions/15681615/how-to-use-capybara-has-text

Finding one of my own old SO questions in the process.  Then I read some of the [capybara docs](http://www.rubydoc.info/github/jnicklas/capybara/master/Capybara/Node/Matchers%3Ahas_content%3F), and then I did more experimenting in the debugger:

```
(byebug) page.has_text? :all, 'is live'
true
(byebug) page.has_content?  'is live'
false
(byebug) page.has_content? :all, 'is live'
true
```

so I conclude that the "is live" is still hidden at this point in the test.  I check that by looking at the rest of the HTML code where I see the class is set to be hidden:

```html
<div class="hidden-sm hidden-xs hidden-md" id="next-event" data-event-time="2014-02-03T07:00:00+00:00" data-event-url="/events/scrum" data-event-name="Scrum">
  <a href="/events/scrum">Scrum</a> is live!
</div>
````

The solution here is NOT to use a different matcher, or change the JavaScript settings, but is it was I said before, the hangout must be taken live, with something like the following:

```
    Given the Hangout for event "Scrum" has been started with details:
      | EventInstance link | http://hangout.test |
      | Started at         | 10:25:00 UTC        |
```

if the hangout is not taken live (by a simulated ping from the google hangout plugin) then it's not going to show up.

Does that make sense?  If it's not making sense, let me know and I can push the change that you need directly into the relevant part of the cucumber test.

The techniques I'm using here with the debugger may not come across very well in text - would be good to go through them with you in a pairing session some time :-)

**Comment by c1505** [comment]
Can we pair on this next week?  I could make myself available Monday 7-9AM EST and will also be available Wednesday starting at 3PM. 

The change to "is live!" happens when a scrum is within the current event duration, not when it actually is live.  The homepage checks if it is live, but not the popdown right below the navbar on other pages.  I checked by manually testing to confirm.

image

**Comment by tansaku** [comment]
@c1505 yes, you're right of course, the "is live" thing does come up during the event duration (bug/feature?) - my guess would then be this is a subtle date issue (I've knocked myself out on those before) - would be great to pair next week.

We have our marketing meeting 7am EST Monday, but we could pair after that - at the minimum 8am-9am EST

**Comment by c1505** [comment]
This is ready for review.  I might refactor the self.next_occurrence method in the future, but I think I would like to extract a class from the Event model first and do some other refactoring.

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -19,9 +18,7 @@
 # in spec/support/ and its subdirectories.
 Dir[Rails.root.join('spec/support/**/*.rb')].each { |f| require f }
 
-# Checks for pending migrations before tests are run.
-# If you are not using ActiveRecord, you can remove this line.
-ActiveRecord::Migration.check_pending! if defined?(ActiveRecord::Migration)
+ActiveRecord::Migration.maintain_test_schema!
```
Comment (tansaku) [inline spec/spec_helper.rb]: so this change seems unrelated - but I guess if everything is green we shouldn't worry - but might be nice to know where this crept in from - was it an explicit change you wanted? If so, is there a blog post or rails doc on the difference?  I see it mentioned here: https://relishapp.com/rspec/rspec-rails/docs/upgrade#pending-migration-checks  looks like a good change to make - ideally in separate PR perhaps, but it sounds like a clear win :-)

**Comment by tansaku** [comment]
great work @c1505 - this is looking really good - apologies for the delay in responding as I was away last week ...

File: spec/spec_helper.rb (hunk 1 above)
Comment (c1505) [inline spec/spec_helper.rb]: It is unrelated.  I meant to comment about it and ask if it makes sense to submit it as a separate pull request.  I am good either way.  I was going to post the link you just did Rails 4.1 and Rspec 3.0 and forward, the change is the default.  I can't think of a reason not to run a pending migration to the test database automatically, but there could be some.

Comment (tansaku) [inline spec/spec_helper.rb]: I agree with you - I think let's merge it in :-)

**Pull request merged**